libparsec/crates/client/tests/unit/certif/rename_realm.rs

```diff
@@ -0,0 +1,255 @@
+async fn server_initial_name_already_exists(env: &TestbedEnv) {
+    let (env, realm_id) = env.customize_with_map(|builder| {
+        builder
+            .new_realm("alice")
+            .then_do_initial_key_rotation()
+            .map(|event| event.realm)
+    });
+    let alice = env.local_device("alice@dev1");
+    let ops = certificates_ops_factory(&env, &alice).await;
+
+    ops.add_certificates_batch(
+        &env.get_common_certificates_signed(),
+        &[],
+        &[],
+        &env.get_realms_certificates_signed(),
+    )
+    .await
+    .unwrap();
+
+    let keys_bundle = env.get_last_realm_keys_bundle(realm_id);
+    let keys_bundle_access = env.get_last_realm_keys_bundle_access_for(realm_id, &alice.user_id());
+    test_register_sequence_of_send_hooks!(
+        &env.discriminant_dir,
+        {
+            move |req: authenticated_cmds::latest::realm_get_keys_bundle::Req| {
+                p_assert_eq!(req.key_index, Some(1));
+                p_assert_eq!(req.realm_id, realm_id);
+
+                authenticated_cmds::latest::realm_get_keys_bundle::Rep::Ok {
+                    key_index: 1,
+                    keys_bundle,
+                    keys_bundle_access,
+                }
+            }
+        },
+        {
+            move |req: authenticated_cmds::latest::realm_rename::Req| {
+                p_assert_eq!(req.initial_name_or_fail, false);
+                authenticated_cmds::latest::realm_rename::Rep::InitialNameAlreadyExists {
+                    last_realm_certificate_timestamp: "2000-01-02T00:00:00Z".parse().unwrap(),
+                }
+            }
+        },
+    );
+
+    let res = ops
+        .rename_realm(realm_id, "wksp1".parse().unwrap())
+        .await
+        .unwrap();
+
+    p_assert_matches!(
+        res,
+        CertificateBasedActionOutcome::RemoteIdempotent { certificate_timestamp }
+        if certificate_timestamp == "2000-01-02T00:00:00Z".parse().unwrap()
+    );
+}
+
+#[parsec_test(testbed = "minimal")]
+async fn invalid_keys_bundle(env: &TestbedEnv) {
+    let (env, realm_id) = env.customize_with_map(|builder| {
+        builder
+            .new_realm("alice")
+            .then_do_initial_key_rotation()
+            .map(|event| event.realm)
+    });
+    let alice = env.local_device("alice@dev1");
+    let ops = certificates_ops_factory(&env, &alice).await;
+
+    ops.add_certificates_batch(
+        &env.get_common_certificates_signed(),
+        &[],
+        &[],
+        &env.get_realms_certificates_signed(),
+    )
+    .await
+    .unwrap();
+
+    test_register_send_hook(&env.discriminant_dir, {
+        move |req: authenticated_cmds::latest::realm_get_keys_bundle::Req| {
+            p_assert_eq!(req.key_index, Some(1));
+            p_assert_eq!(req.realm_id, realm_id);
+
+            authenticated_cmds::latest::realm_get_keys_bundle::Rep::Ok {
+                key_index: 1,
+                keys_bundle: Bytes::from_static(b""),
```

Comment:
parametrize the test so that  invalid `keys_bundle` and invalid `keys_bundle_access` can be both tested separately